pkg/ads/ads.go

```diff
@@ -14,154 +14,146 @@
-		// check if its CIDR.  If so, reject the address unless its /32 CIDR
-		if strings.Contains(address, "/") {
-			if ip, network, err := net.ParseCIDR(address); err != nil {
-				ones, bits := network.Mask.Size()
-				if ones == bits {
-					// its a full mask (e.g., /32). Effectively an IP
-					vips = append(vips, ip)
-				}
-			}
-		} else {
-			if ip := net.ParseIP(address); ip != nil {
-				vips = append(vips, ip)
-			}
-		}
-	}
-
-	return vips
-}
+ import (
+	 "fmt"
+	 "log"
+	 "net"
+	 "strings"
+ 
+	 discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	 "google.golang.org/grpc"
+	 "google.golang.org/grpc/credentials/insecure"
+	 "google.golang.org/protobuf/proto"
+	 "google.golang.org/protobuf/types/known/anypb"
+	 mcp "istio.io/api/mcp/v1alpha1"
+	 networking "istio.io/api/networking/v1alpha3"
+	 "istio.io/istio/pkg/adsc"
+	 "istio.io/istio/pkg/config/schema/collections"
+	 "kmesh.net/kmesh-coredns-plugin/pkg/dns"
+	 "kmesh.net/kmesh-coredns-plugin/pkg/options"
+ )
+ 
+ type AdsController struct {
+	 server    *dns.KmeshDNSServer
+	 adsClient *adsc.ADSC
+	 vip       string
+ }
+ 
+ func NewAdsController(dnsServer *dns.KmeshDNSServer) (*AdsController, error) {
+	 c := &AdsController{
+		 server: dnsServer,
+		 vip:    options.GetConfig().VIP,
+	 }
+	 client, err := adsc.New(options.GetConfig().XDSAddress, &adsc.ADSConfig{
+		 InitialDiscoveryRequests: configInitialRequests(),
+		 ResponseHandler:          c,
+		 Config: adsc.Config{
+			 Workload:  options.GetConfig().ServiceNode,
+			 Namespace: options.GetConfig().ServiceNameSpace,
+			 GrpcOpts: []grpc.DialOption{
+				 // Because we use the custom grpc options for adsc, here we should
+				 // explicitly set transport credentials.
+				 // TODO: maybe we should use the tls settings within ConfigSource
+				 // to secure the connection between istiod and remote xds server.
+				 grpc.WithTransportCredentials(insecure.NewCredentials()),
+			 },
+		 },
+	 })
+	 if err != nil {
+		 log.Fatalf("failed to dial XDS %s %v", "controller.gateway-system.svc.cluster.local:15010", err)
+		 return nil, err
+	 }
+ 
+	 c.adsClient = client
+	 return c, err
+ }
+ 
+ func (c *AdsController) Start(stop <-chan struct{}) error {
+	 if err := c.adsClient.Run(); err != nil {
+		 return err
+	 }
+ 
+	 go func() {
+		 <-stop
+		 c.Stop()
+	 }()
+	 return nil
+ }
+ 
+ func (c *AdsController) Stop() {
+	 c.adsClient.Close()
+ }
+ 
+ func configInitialRequests() []*discovery.DiscoveryRequest {
+	 return []*discovery.DiscoveryRequest{&discovery.DiscoveryRequest{
+		 TypeUrl: collections.ServiceEntry.GroupVersionKind().String(),
+	 }}
+ }
+ 
+ func (c *AdsController) HandleResponse(con *adsc.ADSC, response *discovery.DiscoveryResponse) {
+	 dnsEntries := make(map[string][]net.IP)
+	 for _, r := range response.Resources {
+		 m := &mcp.Resource{}
+		 if err := anypb.UnmarshalTo(r, m, proto.UnmarshalOptions{}); err != nil {
+			 fmt.Println("Error unmarshalling Resource:", err)
```

Comment:
nit:  we should use kmesh logger in the long run, i donot mind using this at first version